### PR TITLE
Implement support of kernel 6.0.x

### LINKED ISSUE #199: Support Linux kernel 6
It's already been [released](https://9to5linux.com/linux-kernel-6-0-officially-released-this-is-whats-new) October 3rd, 2022.
And it's already been used in Fedora 37. So, we have to move on and make `elastio-snap` working with this new major update.

### PR DESCRIPTION
**Pull request by skypodolsky**
It's required for our driver to support kernel 6.0.x. This patch covers this task.

There are bio requests for special purposes, which are configured on the bio allocation by setting a specific bio 
operations field. This field is represented by the enum req_op / enum req_ops. Historically, this enum was called
"req_op" (before v4.10), then it was renamed to "req_ops" (v.4.10+ ... v5.19). Previously, the elastio-snap driver
could handle both of these data types. But the logic worked that way that when in v6.0 that data type has been
renamed back to "req_op", this disabled the support of the "special" bio requests, which caused the memory
corruption.

 - Fixed the problem described above
 - Reverted patch to downgrade the kernel version on Fedora distros
 - Added compat for `bdevname` and `REQ_OP_WRITE_ZEROES` bio request
 - Rewritten `alloc_disk()` logic to make it more straightforward and fixed a refcnt warning

Closes #199

### CODE REVIEW

File: src/configure-tests/feature-tests/bdevname.c [hunk 1]
```diff
@@ -0,0 +1,16 @@
+// SPDX-License-Identifier: GPL-2.0-only
+
+/*
+ * Copyright (C) 2022 Elastio Software Inc.
+ */
+
+// kernel_version > 4.5
```
[the remark this answers is not in the document]
Comment (skypodolsky) [reply src/configure-tests/feature-tests/bdevname.c]: Done.

File: src/elastio-snap.c [hunk 2]
```diff
@@ -3811,7 +3814,11 @@ static int find_orig_fops(struct block_device *bdev, struct block_device_operati
 			*ops = dev->sd_orig_ops;
 			*mrf = dev->sd_orig_mrf;
 			*tracing_ops = tracing_ops_get(dev->sd_tracing_ops);
+#ifdef HAVE_BDEVNAME
 			bdevname(dev->sd_base_dev, bdev_name);
+#else
+			snprintf(bdev_name, sizeof(bdev_name), "%pg", dev->sd_base_dev);
+#endif
```
[the remark this answers is not in the document]
Comment (skypodolsky) [reply src/elastio-snap.c]: Done.

File: src/elastio-snap.c [hunk 3]
```diff
@@ -2881,8 +2881,11 @@ static int bio_needs_cow(struct bio *bio, struct snap_device *dev){
 		return 1;
 	}
 
-#ifdef HAVE_ENUM_REQ_OPF
-//#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,10,0)
+// This is a tricky one:
+// HAVE_ENUM_REQ_OPF: KERNEL_VERSION >= 4.10 && KERNEL_VERSION <= 5.19
+// HAVE_ENUM_REQ_OP: KERNEL_VERSION < 4.10 && KERNEL_VERSION >= 6.0
+#if (defined HAVE_ENUM_REQ_OPF) || \
+	(defined HAVE_ENUM_REQ_OP && LINUX_VERSION_CODE >= KERNEL_VERSION(6,0,0))
```
[the remark this answers is not in the document]
Comment (skypodolsky) [reply src/elastio-snap.c]: Done.

File: src/elastio-snap.c [hunk 4]
```diff
@@ -4208,20 +4223,19 @@ static void __tracer_destroy_snap(struct snap_device *dev){
 #endif
 		if(dev->sd_queue){
 			LOG_DEBUG("freeing request queue");
-#ifdef HAVE_BLK_CLEANUP_QUEUE
+#if defined HAVE_ALLOC_DISK
 			blk_cleanup_queue(dev->sd_queue);
```
[the remark this answers is not in the document]
Comment (skypodolsky) [reply src/elastio-snap.c:4218]: As the logic has been rewritten, we don't need it anymore. This compat has been deleted.

File: src/includes.h [hunk 5]
```diff
@@ -30,6 +30,7 @@
 #include <linux/unistd.h>
 #include <linux/vmalloc.h>
 #include <linux/random.h>
+#include <linux/version.h>
```
Comment (e-kov) [inline src/includes.h:33]: unused, to be removed

File: src/elastio-snap.c [hunk 6]
```diff
@@ -111,6 +111,12 @@ struct path {
 #define elastio_snap_get_mnt(f) (f)->f_path.mnt
 #endif
 
+#ifdef HAVE_BDEVNAME
+#define elastio_snap_bdevname(dev, buf) bdevname(dev, buf)
+#else
+#define elastio_snap_bdevname(dev, buf) snprintf(buf, sizeof(buf), "%pg", dev);
+#endif
```
Comment (e-kov) [inline src/elastio-snap.c:118]: maybe just 
```suggestion
#ifndef HAVE_BDEVNAME
#define bdevname(dev, buf) snprintf(buf, sizeof(buf), "%pg", dev);
#endif
```
?

**Review by anelson — APPROVED** [review]

File: src/includes.h (hunk 5 above)
Comment (skypodolsky) [reply src/includes.h]: Done.

File: src/elastio-snap.c (hunk 6 above)
Comment (skypodolsky) [reply src/elastio-snap.c:118]: I don't think so. I wish the person to understand this is not a default `bdevname`. This should be explicitly mentioned.

**Review by vsazhenyuk-softheme — APPROVED** [review]

File: src/elastio-snap.c (hunk 6 above)
Comment (e-kov) [reply src/elastio-snap.c:118]: ok

**Review by e-kov — APPROVED** [review]

**Pull request merged**